File: 4-Behavioral Patterns/7-Mediator Pattern/Interphone Group Example (1-to-many)/Java/runner/Avengers.java
package runner;

import model.InterphoneGroup;
import model.LeaderUser;
import model.NormalUser;
import model.User;

/**
 * Application that actually uses Mediator Pattern.
 *
 * @author Ziang Lu
 */
public class Avengers {

    /**
     * Main driver.
     * @param args arguments from command line
     */
    public static void main(String[] args) {
        User steve = new LeaderUser("Steve Rogers");
        User tony = new LeaderUser("Tony Stark");
        User bruce = new NormalUser("Bruce Banner");

        InterphoneGroup avengersGroup = new InterphoneGroup();
        steve.joinGroup(avengersGroup);
        tony.joinGroup(avengersGroup);
        bruce.joinGroup(avengersGroup);

        tony.sendMessage("Incoming enemy!");

        /*
         * Output:
         * Tony Stark sends a message: Incoming enemy!
         * Steve Rogers received a message from Tony Stark: Incoming enemy!
         * Bruce Banner received a message from Tony Stark: Incoming enemy!
         */
    }

}

File: 4-Behavioral Patterns/7-Mediator Pattern/Interphone Group Example (1-to-many)/Python/avengers.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Application that actually uses Mediator Pattern.
"""

__author__ = 'Ziang Lu'

from model import InterphoneGroup, LeaderUser, NormalUser


def main():
    steve = LeaderUser('Steve')
    tony = LeaderUser('Tony')
    bruce = NormalUser('Bruce')

    group = InterphoneGroup()
    steve.join_group(group)
    tony.join_group(group)
    bruce.join_group(group)

    tony.send_message('Incoming enemy!')


if __name__ == '__main__':
    main()

# Output:
# Tony sends a message: Incoming enemy!
# Steve received a message from Tony: Incoming enemy!
# Bruce received a message from Tony: Incoming enemy!
